src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs

```diff
@@ -19,21 +19,40 @@ namespace Microsoft.CodeAnalysis.Diagnostics
     /// </summary>
     internal partial class AnalyzerManager
     {
-        /// <summary>
-        /// Gets the default instance of the AnalyzerManager for the lifetime of the analyzer host process.
-        /// </summary>
-        public static readonly AnalyzerManager Instance = new AnalyzerManager();
+        private readonly object _gate = new object();
 
         // This cache stores the analyzer execution context per-analyzer (i.e. registered actions, supported descriptors, etc.).
-        private readonly ConditionalWeakTable<DiagnosticAnalyzer, AnalyzerExecutionContext> _analyzerExecutionContextMap =
-            new ConditionalWeakTable<DiagnosticAnalyzer, AnalyzerExecutionContext>();
+        private readonly ImmutableDictionary<DiagnosticAnalyzer, AnalyzerExecutionContext> _analyzerExecutionContextMap;
+
+        public AnalyzerManager(ImmutableArray<DiagnosticAnalyzer> analyzers)
+        {
+            _analyzerExecutionContextMap = CreateAnalyzerExecutionContextMap(analyzers);
+        }
+
+        public AnalyzerManager(DiagnosticAnalyzer analyzer)
+        {
+            _analyzerExecutionContextMap = CreateAnalyzerExecutionContextMap(SpecializedCollections.SingletonEnumerable(analyzer));
+        }
+
+        private ImmutableDictionary<DiagnosticAnalyzer, AnalyzerExecutionContext> CreateAnalyzerExecutionContextMap(IEnumerable<DiagnosticAnalyzer> analyzers)
```

Comment:
Hum, the singleton case may fit better with IEnumerable (than ImmutableArray). Ignore that comment.

---
In reply to: [121486170](https://github.com/dotnet/roslyn/pull/20109#discussion_r121486170) [](ancestors = 121486170)